### PR TITLE
Adding nested sampling with NestedSamplers.jl

### PR DESCRIPTION
**Pull request by WillyHK**

### CODE REVIEW

**Comment by oschulz** [comment]
Nice going, guys!

File: src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl [hunk 1]
```diff
@@ -5,18 +5,57 @@ using .MCMCChains: Chains
 include("tns_bounds.jl")
 include("tns_proposals.jl")
 
+"""
+    struct TuringNestedSamplers <: AbstractUltraNestAlgorithm
```
Comment (Cornelius-G) [inline src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl:9]: I guess the `AbstractUltraNestAlgorithm` here is a copy paste error :wink:

File: src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl [hunk 2]
```diff
@@ -5,18 +5,57 @@ using .MCMCChains: Chains
...
     trafo::TR = PriorToUniform()
 
-    num_live_points::Int = 1000                         # the number of live-points
-    bound::TNS_Bound = TNS_MultiEllipsoidBound()        # volume around the live-points
-    proposal::TNS_Proposal = TNS_AutoProposal()         # algorithm to choose new live-points
-    enlarge::Float64 = 1.25                             # scale-factor for the volume
-    # update_interval::Float64 =                        # not sure about how this works yet
-    min_ncall::Int64 = 2*num_live_points                # number of iterations before the first bound will be fit
-    min_eff::Float64 = 0.10                             # efficiency before fitting the first bound
+    "Number of live-points."
```
Comment (Cornelius-G) [inline src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl:37]: Also, I don't think you can just put those stings in here. What you had before was ok, I guess.

File: src/samplers/nested_sampling/turing_nestedSamplers/tns_proposals.jl [hunk 3]
```diff
@@ -0,0 +1,66 @@
+# Here are the definitions of the NestedSamplers proposals, which discribe the algorithm to choose new live-points
+export TNS_Uniformly
```
Comment (oschulz) [inline src/samplers/nested_sampling/turing_nestedSamplers/tns_proposals.jl:2]: Let's not export these until they are documented properly.

File: src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    max_iters = Inf
+    max_ncalls = Inf
+    maxlogl = Inf
+end
+export TuringNestedSamplers
+
+
+############################################################################################################
+# Here is the bat_sample implementation for the NestedSamplers package
+############################################################################################################
```
Comment (oschulz) [inline src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl:69]: BAT.jl doesn't use "headline" comments like this.

File: src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    nsamples = size(chain.value.data,1)
+    samples = [chain.value.data[i, 1:end-1] for i in 1:nsamples]                            # the other ones (between 1 and end-1) are the samples
+    logvals = map(logdensityof(density), samples)                                           # posterior values of the samples
+    samples_trafo = vs.(BAT.DensitySampleVector(samples, logvals, weight = weights))
+    samples_notrafo = inv(trafo).(samples_trafo)                                            # Here the samples are retransformed
+
+    return (                                                                                # possibly there are more informations which could been returned
+        result = samples_notrafo, result_trafo = samples_trafo, trafo = trafo,
+    )
+end
```
Comment (oschulz) [inline src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl:108]: File should end with a newline.

File: src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+function bat_sample_impl(rng::AbstractRNG, target::AnyDensityLike, algorithm::TuringNestedSamplers)
+    
+    density_notrafo = convert(AbstractDensity, target)
+    shaped_density, trafo = bat_transform(algorithm.trafo, density_notrafo)                 # BAT prior transformation
+    vs = varshape(shaped_density)
+    density = unshaped(shaped_density)
+    dims = totalndof(vs)
+
+    # This function is used for likelihood calculations by NestedSamplers. It has to be a fuction of only x. It includes the prior.
+    function tns_posterior(x)
```
Comment (oschulz) [inline src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl:79]: Use `tns_posterior = logdensityof(density)` instead.

File: src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        return BAT.eval_logval_unchecked(density,x)
+    end
+
+    tns_prior_trafo(x) = x                                                                  # identity, because the we use the BAT version of prior transformation instead
+    model = NestedModel(tns_posterior, tns_prior_trafo);
```
Comment (oschulz) [inline src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl:84]: Just pass `identity` as `tns_prior_trafo`.

File: src/BAT.jl [hunk 8]
```diff
@@ -93,6 +93,9 @@ _plots_module() = _PLOTS_MODULE[]
 function __init__()
     @require Cuba = "8a292aeb-7a57-582c-b821-06e4c11590b1" include("integration/cuba_integration.jl")
     @require HDF5 = "f67ccb44-e63f-5c2f-98bd-6dc0ccc4ba2f" include("io/hdf5_specific.jl")
+    @require NestedSamplers = "41ceaf6f-1696-4a54-9b49-2e7a9ec3782e" begin
+        @require MCMCChains = "c7f686f2-ff18-58e9-bc7b-31028e88f75d" include("samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl") # loads if both packages there
```
Comment (oschulz) [inline src/BAT.jl:97]: Do we need to require `MCMCChains` (do we uses types of functions in it explicitly)?

File: src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+    weights = chain.value.data[:, end]                                                      # the last elements of the vectors are the weights
+    nsamples = size(chain.value.data,1)
+    samples = [chain.value.data[i, 1:end-1] for i in 1:nsamples]                            # the other ones (between 1 and end-1) are the samples
+    logvals = map(logdensityof(density), samples)                                           # posterior values of the samples
+    samples_trafo = vs.(BAT.DensitySampleVector(samples, logvals, weight = weights))
+    samples_notrafo = inv(trafo).(samples_trafo)                                            # Here the samples are retransformed
+
+    return (                                                                                # possibly there are more informations which could been returned
+        result = samples_notrafo, result_trafo = samples_trafo, trafo = trafo,
```
Comment (oschulz) [inline src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl:106]: Should return the integral estimate as well, it's one of the main advantages of nested sampling, after all.

**Review by oschulz — CHANGES_REQUESTED** [review]

**Comment by oschulz** [comment]
Nice! I added a review with a few comments. And there should be some tests, at least, before we merge it. :-)

**Comment by oschulz** [comment]
Regarding ESS estimation, this may be of interest, it's fairly recent: https://arxiv.org/abs/1602.03572

**Comment by WillyHK** [comment]
I took another quick look and everything seems to work, have run the tests several times and there was no problem.

I think the code should be possible to merge

**Comment by oschulz** [comment]
There's a merge conflict regarding your additions to "Project.toml" - test dependencies are now in "test/Project.toml", not in the main "Project.toml" anymore. Could you change the PR accordingly?

**Comment by WillyHK** [comment]
Oh, sry for that, hopefully it's fine now @oschulz

File: src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl [hunk 10]
```diff
@@ -0,0 +1,98 @@
+using .NestedSamplers               # used in this file, because this file ist only load if it is required
+
+include("tns_bounds.jl")
+include("tns_proposals.jl")
+
+"""
+    struct TuringNestedSamplers <: AbstractSamplingAlgorithm
```
Comment (oschulz) [inline src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl:7]: Let's rename this to `EllipsoidalNestedSampling`

File: test/samplers/nested_sampling/test_turing_nestedSamplers.jl [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+using BAT
+using Test
+
+using Random, StatsBase, Distributions
+using HypothesisTests
+
+import NestedSamplers
+
+
+@testset "test_turing_nestedSamplers" begin
```
Comment (oschulz) [inline test/samplers/nested_sampling/test_turing_nestedSamplers.jl:12]: Should be `@testset "ellipsoidal_nested_sampling"`

File: test/samplers/nested_sampling/test_turing_nestedSamplers.jl [hunk 12]
```diff
@@ -0,0 +1,44 @@
+# This file is a part of BAT.jl, licensed under the MIT License (MIT).
```
Comment (oschulz) [inline test/samplers/nested_sampling/test_turing_nestedSamplers.jl:1]: Let's name this file "test_ellipsoidal_nested_sampling.jl"

File: src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl [hunk 13]
```diff
@@ -0,0 +1,98 @@
+using .NestedSamplers               # used in this file, because this file ist only load if it is required
```
Comment (oschulz) [inline src/samplers/nested_sampling/turing_nestedSamplers/turing_nestedSamplers.jl:1]: Let's name this file "ellipsoidal_nested_sampling.jl"

**Review by oschulz — CHANGES_REQUESTED** [review]

File: src/samplers/nested_sampling/ellipsoidal_nested_sampling/ens_bounds.jl [hunk 14]
```diff
@@ -0,0 +1,49 @@
+# This file is a part of BAT.jl, licensed under the MIT License (MIT).
+
+"""
+    abstract type ENS_Bound 
+
+Abstract type for the bounds of the sampling region used by EllipsoidalNestedSampling.
+"""
+abstract type ENS_Bound end
```
Comment (oschulz) [inline src/samplers/nested_sampling/ellipsoidal_nested_sampling/ens_bounds.jl:8]: Type names should not contain underscores, just use `ENSBounds` and so on.

File: src/samplers/nested_sampling/ellipsoidal_nested_sampling/ens_bounds.jl [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+"""
+    struct ENS_NoBounds <: ENS_Bound
+
+No bounds means that the whole volume from the unit Cube is used to find new points.
+"""
+struct ENS_NoBounds <: ENS_Bound end
+export ENS_NoBounds
```
Comment (oschulz) [inline src/samplers/nested_sampling/ellipsoidal_nested_sampling/ens_bounds.jl:16]: Let's not export these for now. So no `export`, and use `struct BAT.ENSNoBounds` in the docstring. Also all docstrings should contain `*Experimental feature, not part of stable public API.*` for now.

File: src/samplers/nested_sampling/ellipsoidal_nested_sampling/ens_proposals.jl [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+     ndims < 10: Proposals.Uniform, 
+     10 ≤ ndims ≤ 20: Proposals.RWalk, 
+     ndims > 20: Proposals.Slice
+"""
+struct ENS_AutoProposal <: ENS_Proposal end
+export ENS_AutoProposal
+
+"""
+    struct ENS_RandomWalk <: ENS_Proposal
+
```
Comment (oschulz) [inline src/samplers/nested_sampling/ellipsoidal_nested_sampling/ens_proposals.jl:31]: For `@with_kw` structs, add the fields to the docstring by including

```julia
Constructors:

* ```$(FUNCTIONNAME)(; fields...)```

Fields:

$(TYPEDFIELDS)
```

See ultranest.jl, etc.

File: src/samplers/nested_sampling/ellipsoidal_nested_sampling/ens_proposals.jl [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+New live point is proposed by using a random walk away from an existing live point.
+"""
+@with_kw struct ENS_RandomWalk <: ENS_Proposal
+    ratio::Float64 = 0.5
```
Comment (oschulz) [inline src/samplers/nested_sampling/ellipsoidal_nested_sampling/ens_proposals.jl:35]: Maybe document the semantics of the fields (see ultranest.jl for examples).

**Review by oschulz — CHANGES_REQUESTED** [review]

**Review by oschulz — APPROVED** [review]

**Comment by oschulz** [comment]
Looks good to merge, as soon as the test failures are gone.

**Comment by WillyHK** [comment]
I have found an error message in the full log, which seems to be the reason for the fail:

 "Warning: Error requiring` NestedSamplers` from `BAT`"
 "UndefVarError: EllipsoidalNestedSampling not defined"

I have now checked the code again, but actually found no reason for it.

Could the problem maybe be that the tests are running in an environment (on a Gitserver?), where NestedSamplers.jl was not added? 
(I think that would be necessary for the @require statement so that the code is loaded)

Please apologize if this is a wrong thought, may you have another idea why the error occurs @oschulz

**Comment by oschulz** [comment]
We still have this `invalid keyword argument syntax "prop.ratio"` error on Julia v1.3 x64 - any idea why?

**Comment by WillyHK** [comment]
Not really.
The only codeline, where prop.ratio is used, is in the return function for ENSRandomWalk in ens_proposals, but there it looks fine to me.

**Comment by oschulz** [comment]
I'll re-run the tests to see if it's a sporadic failure.

**Comment by oschulz** [comment]
Nope, seems to always fail on Julia 1.3 ubuntu x64. We should figure this out before we merge.

**Comment by oschulz** [comment]
@WillyHK could you try running the tests locally (`Pkg.test("BAT")`), on a 64-bit Linux system with Julia 1.3 ?

**Comment by WillyHK** [comment]
With julia 1.6 it runs first try, locally at Ubuntu x64, I will later download version 1.3 and test it again

**Comment by WillyHK** [comment]
Update: with julia version 1.3 the known error `invalid keyword argument syntax "prop.ratio"` occured locally too, on the same system it worked with julia 1.6. So it seems to be a problem with the old julia version

**Comment by WillyHK** [comment]
Update: I try a few more tries with different versions of julia, the result is:

With julia 1.3 or 1.4 the test fails with the error in the upper comment
With julia 1.5 or 1.6 all tests are succesfull

**Comment by oschulz** [comment]
> With julia 1.3 or 1.4 the test fails with the error in the upper comment

If that happens locally too, can you and @Cornelius-G try to figure out why it happens?

**Comment by WillyHK** [comment]
So, now locally all test passed succesfully for julia 1.3

Thank you for giving me the hint, that the julia version could be the reason for the testfail @oschulz 

the problem was that julia changed the handling of different syntaxes for optional keywords over the versions, which i didn't know.

**Comment by oschulz** [comment]
> Julia changed the handling of different syntaxes for optional keywords over the versions

Oh, that's where it was - of course, sorry, I should have seen that. Ok, let's see if tests pass now. :-)

**Comment by oschulz** [comment]
Thanks!!

**Pull request merged**